### PR TITLE
[CUDA] Add CUDA_ARCHITECTURES to fix CMake warnings (#3754)

### PR DESCRIPTION
**Pull request by RobinDong**
Set property `CUDA_ARCHITECTURES` would fix the warnings reported by CMake 3.20.x

### CODE REVIEW

**Review by jameslamb — COMMENTED** [review]
Thanks very much! I checked and it does seem that these changes will resolve the warnings with newer versions of `CMake`.

* CUDA 11.2.2 build on latest `master` (warnings from #3754): https://github.com/microsoft/LightGBM/runs/2540826588
* CUDA 11.2.2 build for this PR (no warnings): https://github.com/microsoft/LightGBM/pull/4268/checks?check_run_id=2541179844

<hr>

Those CI jobs test with the latest version of `CMake`, which is 3.20.2 as of today.

```text
Get:1 https://apt.kitware.com/ubuntu focal/main amd64 cmake-data all 3.20.2-0kitware1ubuntu20.04.1 [1807 kB]
Get:2 https://apt.kitware.com/ubuntu focal/main amd64 cmake amd64 3.20.2-0kitware1ubuntu20.04.1 [10.5 MB]
```

Have you tested these changes against older versions of `CMake`? LightGBM currently supported `CMake>=3.16`  for CUDA builds.

https://github.com/microsoft/LightGBM/blob/c1d2dbe2c546e3d0c6f9c66ef5dcef6428541df2/CMakeLists.txt#L27-L28

I'm just leaving a "comment" review because I'd like to be sure this change doesn't break compatibility with older `CMake` versions. If you haven't tested that and aren't able to easily test it, I can try in the next day or two.

**Comment by RobinDong** [comment]
> Thanks very much! I checked and it does seem that these changes will resolve the warnings with newer versions of `CMake`.
> 
> * CUDA 11.2.2 build on latest `master` (warnings from #3754): https://github.com/microsoft/LightGBM/runs/2540826588
> * CUDA 11.2.2 build for this PR (no warnings): https://github.com/microsoft/LightGBM/pull/4268/checks?check_run_id=2541179844
> 
> Those CI jobs test with the latest version of `CMake`, which is 3.20.2 as of today.
> 
> ```
> Get:1 https://apt.kitware.com/ubuntu focal/main amd64 cmake-data all 3.20.2-0kitware1ubuntu20.04.1 [1807 kB]
> Get:2 https://apt.kitware.com/ubuntu focal/main amd64 cmake amd64 3.20.2-0kitware1ubuntu20.04.1 [10.5 MB]
> ```
> 
> Have you tested these changes against older versions of `CMake`? LightGBM currently supported `CMake>=3.16` for CUDA builds.
> 
> https://github.com/microsoft/LightGBM/blob/c1d2dbe2c546e3d0c6f9c66ef5dcef6428541df2/CMakeLists.txt#L27-L28
> 
> I'm just leaving a "comment" review because I'd like to be sure this change doesn't break compatibility with older `CMake` versions. If you haven't tested that and aren't able to easily test it, I can try in the next day or two.

Hi James,

Thank you for your review. 
I have tested this patch for CMake-3.16.9 and CMake-3.19.8 by using `cmake -DCMAKE_CUDA_COMPILER=/usr/local/cuda/bin/nvcc -DUSE_CUDA=ON ..` and they all work well.

File: CMakeLists.txt [hunk 1]
```diff
@@ -173,6 +173,10 @@ if(USE_CUDA)
         list(APPEND CUDA_ARCHS "7.5")
     endif()
     list(POP_BACK CUDA_ARCHS CUDA_LAST_SUPPORTED_ARCH)
+    string(SUBSTRING ${CUDA_LAST_SUPPORTED_ARCH} 0 1 MAJOR_VERSION)
+    string(SUBSTRING ${CUDA_LAST_SUPPORTED_ARCH} 2 1 MINOR_VERSION)
+    string(CONCAT CUDA_ARCH_VERSION ${MAJOR_VERSION} ${MINOR_VERSION})
```
Comment (StrikerRUS) [inline CMakeLists.txt:178]: I guess this will fail once CUDA ARCH become 2-digit.
Maybe we can simply replace `.` with empty string?
Or use some workarounds for splitting a sting: https://gitlab.kitware.com/cmake/cmake/-/issues/21970#note_952813.

File: CMakeLists.txt [hunk 2]
```diff
@@ -209,7 +213,7 @@ if(USE_CUDA)
 
     function(add_histogram hsize hname hadd hconst hdir)
       add_library(histo${hsize}${hname} OBJECT src/treelearner/kernels/histogram${hsize}.cu)
-      set_target_properties(histo${hsize}${hname} PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
+      set_target_properties(histo${hsize}${hname} PROPERTIES CUDA_SEPARABLE_COMPILATION ON CUDA_ARCHITECTURES ${CUDA_ARCH_VERSION})
```
Comment (StrikerRUS) [inline CMakeLists.txt:216]: For better readability.
```suggestion
      set_target_properties(histo${hsize}${hname} PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
      set_target_properties(histo${hsize}${hname} PROPERTIES CUDA_ARCHITECTURES ${CUDA_ARCH_VERSION})
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -430,12 +434,12 @@ if(__INTEGRATE_OPENCL)
 endif()
 
 if(USE_CUDA)
-  set_target_properties(lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON)
+  set_target_properties(lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON CUDA_ARCHITECTURES ${CUDA_ARCH_VERSION})
```
Comment (StrikerRUS) [inline CMakeLists.txt:437]: For better readability.
```suggestion
  set_target_properties(lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON)
  set_target_properties(lightgbm PROPERTIES CUDA_ARCHITECTURES ${CUDA_ARCH_VERSION})
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -430,12 +434,12 @@ if(__INTEGRATE_OPENCL)
...
   TARGET_LINK_LIBRARIES(
     lightgbm
     ${histograms}
   )
-  set_target_properties(_lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON)
+  set_target_properties(_lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON CUDA_ARCHITECTURES ${CUDA_ARCH_VERSION})
```
Comment (StrikerRUS) [inline CMakeLists.txt:442]: For better readability.
```suggestion
  set_target_properties(_lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON)
  set_target_properties(_lightgbm PROPERTIES CUDA_ARCHITECTURES ${CUDA_ARCH_VERSION})
```

File: CMakeLists.txt (hunk 1 above)
Comment (StrikerRUS) [reply CMakeLists.txt:178]: Could you please explain why do we set only the latest CUDA architecture while XGBoost sets all supported archs?
https://github.com/dmlc/xgboost/blob/e41619b1fc6152d4392896c6caf9a0dc960c5558/cmake/Utils.cmake#L104-L122
https://github.com/dmlc/xgboost/blob/e41619b1fc6152d4392896c6caf9a0dc960c5558/cmake/Utils.cmake#L142-L144

**Review by StrikerRUS — COMMENTED** [review]
@RobinDong Thanks a lot for picking this issue up!
I have some comments and question below:

File: CMakeLists.txt (hunk 1 above)
Comment (RobinDong) [reply CMakeLists.txt]: > Maybe we can simply replace `.` with empty string?
Good idea!

**Comment by jameslamb** [comment]
I've updated the title of this pull request to be a bit more explicit. PR titles become changelog entries in this project (see, for example, https://github.com/microsoft/LightGBM/releases/tag/v3.2.1), and I want to make it easier for anyone looking for CUDA-related changes in the next release to find them.

File: CMakeLists.txt (hunk 1 above)
Comment (StrikerRUS) [reply CMakeLists.txt]: > Sets all supported archs could make compiled code run at all archs of GPUs no matter old ones (Kepler, Maxwell) or new ones (Turing, Ampere)

Thanks a lot of the explanation!

File: CMakeLists.txt [hunk 5]
```diff
@@ -171,8 +171,16 @@ if(USE_CUDA)
     set(CUDA_ARCHS "6.0" "6.1" "6.2" "7.0")
     if(CUDA_VERSION VERSION_GREATER_EQUAL "10.0")
         list(APPEND CUDA_ARCHS "7.5")
+    elseif(CUDA_VERSION VERSION_GREATER_EQUAL "11.0")
+        list(APPEND CUDA_ARCHS "8.0")
```
Comment (StrikerRUS) [inline CMakeLists.txt:175]: Could you please extract this part into a separate PR? This is not related to fixing CMake warnings.
Also, in new PR `8.6` can be appended, right?

File: CMakeLists.txt [hunk 6]
```diff
@@ -171,8 +171,16 @@ if(USE_CUDA)
...
     endif()
     list(POP_BACK CUDA_ARCHS CUDA_LAST_SUPPORTED_ARCH)
+    foreach(arch ${CUDA_ARCHS})
+      string(REPLACE "." "" ver ${arch})
+      set(CUDA_ARCH_VERSION "${ver}-real;${ver}-virtual;${CUDA_ARCH_VERSION}")
+    endforeach()
```
Comment (StrikerRUS) [inline CMakeLists.txt:181]: Can we simplify the whole loop to two lines according to the https://cmake.org/cmake/help/latest/policy/CMP0104.html#examples?

```
# convert CUDA_ARCHS to string (let say CUDA_ARCHS_STRING) and remove from it all dots (.)
set_property(TARGET tgt PROPERTY CUDA_ARCHITECTURES CUDA_ARCHS_STRING)
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 5 above)
Comment (RobinDong) [reply CMakeLists.txt:175]: You are right. I will remove it in this PR

File: CMakeLists.txt [hunk 7]
```diff
@@ -172,6 +172,9 @@ if(USE_CUDA)
     if(CUDA_VERSION VERSION_GREATER_EQUAL "10.0")
         list(APPEND CUDA_ARCHS "7.5")
     endif()
+    list(JOIN CUDA_ARCHS "\;" CUDA_ARCHS_STRING)
```
Comment (StrikerRUS) [inline CMakeLists.txt:175]: In example they use a whitespace as a separator.
```
set_property(TARGET tgt PROPERTY CUDA_ARCHITECTURES 35 50 72)
```
I'm not a CMake expert, so not sure whether it matters. Do you feel confident about using `;` here?

My dirty guess is that one can pass any values to this property in CMake 3.20.2 version (which is one we use at our CI) just with the aim to shut up the warning. This is **not** what we want to achieve in this PR.

```
set_property(TARGET tgt PROPERTY CUDA_ARCHITECTURES foo bar)
```
> CMake version 3.20.2 warns when the policy is not set and uses `OLD` behavior.

File: CMakeLists.txt [hunk 8]
```diff
@@ -210,6 +213,7 @@ if(USE_CUDA)
     function(add_histogram hsize hname hadd hconst hdir)
       add_library(histo${hsize}${hname} OBJECT src/treelearner/kernels/histogram${hsize}.cu)
       set_target_properties(histo${hsize}${hname} PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
+      set_target_properties(histo${hsize}${hname} PROPERTIES CUDA_ARCHITECTURES "${CUDA_ARCHS_STRING}")
```
Comment (StrikerRUS) [inline CMakeLists.txt:216]: Do you have any thought about how is this property compatible with following two lines
https://github.com/microsoft/LightGBM/blob/adf36d766afb6261276d0624f4f5b216d51880c8/CMakeLists.txt#L177
https://github.com/microsoft/LightGBM/blob/adf36d766afb6261276d0624f4f5b216d51880c8/CMakeLists.txt#L179
where we manually append target architectures taken from `CUDA_SELECT_NVCC_ARCH_FLAGS` function to compiler flags?
https://cmake.org/cmake/help/latest/module/FindCUDA.html#commands

Comment (StrikerRUS) [reply CMakeLists.txt:216]: Maybe it will be better to simply set `CUDA_ARCHITECTURES` to `OFF` as it is shown in the last example?
https://cmake.org/cmake/help/latest/policy/CMP0104.html#examples
I believe by doing so we will preserve the old behavior (setting architectures manually in the compiler flags) and in in consequence compatibility with CMake < 3.18 and make the warning disappear in CMake >= 3.18. Please let me know WDYT about this.

File: CMakeLists.txt (hunk 5 above)
Comment (StrikerRUS) [reply CMakeLists.txt]: Thank you! Please feel free to provide a PR with new architectures added.

File: CMakeLists.txt (hunk 8 above)
Comment (RobinDong) [reply CMakeLists.txt:216]: Hmmm, I agree with that. Since already manually append target arch into CMAKE_CUDA_FLAGS, we can just turn off `CUDA_ARCHITECTURES` to keep this change the simplest.

**Comment by StrikerRUS** [comment]
@jameslamb 
> If you haven't tested that and aren't able to easily test it, I can try in the next day or two.

You've said you have an easy access to Linux with CUDA machine. If that's still true, could you please check this PR with CMake 3.16? I want to be sure these changes don't make old CMake panic with unrecognized target property.
https://github.com/microsoft/LightGBM/pull/4268#discussion_r631840059

**Comment by jameslamb** [comment]
>  If that's still true, could you please check this PR with CMake 3.16?

yep, definitely! I'll try soon and let you know what I find.

**Comment by jameslamb** [comment]
Compilation seemed to succeed and I don't see any warnings! Using CUDA 10.2 and CMake 3.16.9 (the latest version in the 3.16 series, as far as I could tell from https://cmake.org/cmake/help/latest/release/3.16.html).

<details><summary>output of nvidia-smi</summary>

```text
Fri May 14 20:01:32 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.100      Driver Version: 440.100      CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 207...  Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   51C    P8    10W /  N/A |    286MiB /  7982MiB |     12%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

</details>

I used `nvidia-docker` to test this. Ran this command:

```shell
nvidia-docker run \
    --env DEBIAN_FRONTEND="noninteractive" \
    -it nvidia/cuda:10.2-devel-ubuntu18.04 \
    /bin/bash
```

Then this script inside the container:

<details><summary>testing scripts</summary>

```shell
apt-get update
apt-get install \
    -y \
    software-properties-common \
    apt-transport-https \
    curl \
    git \
    wget \
    libcurl4-openssl-dev \
    default-jdk-headless \
    libssl-dev \
    libxml2-dev \
    libboost-dev \
    libboost-system-dev \
    libboost-filesystem-dev \
    clinfo

# see install instructions at https://cmake.org/install/
mkdir -p /tmp/cmake-install
pushd /tmp/cmake-install

    CMAKE_VERSION=3.16.9
    CMAKE_TARBALL=cmake-${CMAKE_VERSION}.tar.gz

    curl -OL \
        https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION}/${CMAKE_TARBALL}
    tar -xvzf ${CMAKE_TARBALL}
    rm -rf ./${CMAKE_TARBALL}
    cd cmake-${CMAKE_VERSION}
    ./bootstrap
    make
    make install
popd

# check CMake version
cmake --version

# get LightGBM from PR 4268
git clone --recursive https://github.com/RobinDong/LightGBM.git --branch master

cd LightGBM
mkdir ./build
cd ./build
cmake \
    -DCMAKE_CUDA_COMPILER=/usr/local/cuda/bin/nvcc \
    -DUSE_CUDA=ON \
    ..

make _lightgbm -j4
```

</details>

See the logs below.

<details><summary>logs from 'cmake ..'</summary>

ran the following

```shell
cmake \
    -DCMAKE_CUDA_COMPILER=/usr/local/cuda/bin/nvcc \
    -DUSE_CUDA=ON \
    ..
```

which resulted in these logs

```text
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The CUDA compiler identification is NVIDIA 10.2.89
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc -- works
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr/local/cuda (found suitable version "10.2", minimum required is "9.0") 
-- CMAKE_CUDA_FLAGS: -Xcompiler=-fopenmp -Xcompiler=-fPIC -Xcompiler=-Wall -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_62,code=sm_62 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_75,code=compute_75 -O3 -lineinfo
-- ALLFEATS_DEFINES: -DPOWER_FEATURE_WORKGROUPS=12;-DUSE_CONSTANT_BUF=0;-DENABLE_ALL_FEATURES
-- FULLDATA_DEFINES: -DPOWER_FEATURE_WORKGROUPS=12;-DUSE_CONSTANT_BUF=0;-DENABLE_ALL_FEATURES;-DIGNORE_INDICES
-- Performing Test MM_PREFETCH
-- Performing Test MM_PREFETCH - Success
-- Using _mm_prefetch
-- Performing Test MM_MALLOC
-- Performing Test MM_MALLOC - Success
-- Using _mm_malloc
-- Configuring done
-- Generating done
-- Build files have been written to: /root/LightGBM/build
```

</details>

<details><summary>logs from 'make _lightgbm'</summary>

ran the following

```shell
make _lightgbm -j2
```

which resulted in the following logs

```text
Scanning dependencies of target histo_16_64_256_sp
Scanning dependencies of target histo_16_64_256-fulldata_sp_const
[  2%] Building CUDA object CMakeFiles/histo_16_64_256_sp.dir/src/treelearner/kernels/histogram_16_64_256.cu.o
[  5%] Building CUDA object CMakeFiles/histo_16_64_256-fulldata_sp_const.dir/src/treelearner/kernels/histogram_16_64_256.cu.o
[  5%] Built target histo_16_64_256-fulldata_sp_const
Scanning dependencies of target histo_16_64_256_sp_const
[  7%] Building CUDA object CMakeFiles/histo_16_64_256_sp_const.dir/src/treelearner/kernels/histogram_16_64_256.cu.o
[  7%] Built target histo_16_64_256_sp
Scanning dependencies of target histo_16_64_256-fulldata_sp
[ 10%] Building CUDA object CMakeFiles/histo_16_64_256-fulldata_sp.dir/src/treelearner/kernels/histogram_16_64_256.cu.o
[ 10%] Built target histo_16_64_256_sp_const
[ 10%] Built target histo_16_64_256-fulldata_sp
Scanning dependencies of target _lightgbm
[ 15%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/boosting.cpp.o
[ 15%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt.cpp.o
[ 17%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt_model_text.cpp.o
[ 20%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt_prediction.cpp.o
[ 22%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/prediction_early_stop.cpp.o
[ 25%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/bin.cpp.o
[ 27%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/config.cpp.o
[ 30%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/config_auto.cpp.o
[ 32%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/dataset.cpp.o
[ 35%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/dataset_loader.cpp.o
[ 37%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/file_io.cpp.o
[ 40%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/json11.cpp.o
[ 42%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/metadata.cpp.o
[ 45%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/parser.cpp.o
[ 47%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/train_share_states.cpp.o
[ 50%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/tree.cpp.o
[ 52%] Building CXX object CMakeFiles/_lightgbm.dir/src/metric/dcg_calculator.cpp.o
[ 55%] Building CXX object CMakeFiles/_lightgbm.dir/src/metric/metric.cpp.o
[ 57%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/ifaddrs_patch.cpp.o
[ 60%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linker_topo.cpp.o
[ 62%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linkers_mpi.cpp.o
[ 65%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linkers_socket.cpp.o
[ 67%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/network.cpp.o
[ 70%] Building CXX object CMakeFiles/_lightgbm.dir/src/objective/objective_function.cpp.o
[ 72%] Building CUDA object CMakeFiles/_lightgbm.dir/src/treelearner/cuda_kernel_launcher.cu.o
[ 75%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/cuda_tree_learner.cpp.o
[ 77%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o
[ 80%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/feature_parallel_tree_learner.cpp.o
[ 82%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/gpu_tree_learner.cpp.o
[ 85%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/linear_tree_learner.cpp.o
[ 87%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/serial_tree_learner.cpp.o
[ 90%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/tree_learner.cpp.o
[ 92%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/voting_parallel_tree_learner.cpp.o
[ 95%] Building CXX object CMakeFiles/_lightgbm.dir/src/c_api.cpp.o
[ 97%] Linking CUDA device code CMakeFiles/_lightgbm.dir/cmake_device_link.o
[100%] Linking CXX shared library ../lib_lightgbm.so
[100%] Built target _lightgbm
```

</details>

Now that I have this script working, it would be fairly easy and quick to test other CMake versions, so please let me know if there are other versions you'd like me to check.

**Review by StrikerRUS — APPROVED** [review]
@jameslamb Thank you very much for verifying with CMake 3.16!

@RobinDong Many thanks for your great help! We will really appreciate your further contributions to CMake and CUDA things.

**Comment by StrikerRUS** [comment]
@RobinDong Please update to the latest `master` branch to fix Dask environmental issues we fixed recently and we are ready to merge.
